### ISSUE TITLE
Make teacher reg custom fields more flexible and accessible

### ISSUE THREAD

**Issue by willgearty** [learning-unlimited/ESP-Website#2707]
It's possible to add custom fields to the class registration form using the `teacherreg_custom_forms` tag and a hardcoded form [here](https://github.com/learning-unlimited/ESP-Website/blob/main/esp/esp/program/modules/forms/teacherreg_custom.py). But this functionality shouldn't require chapters to have to hard code form fields to customize the class reg form.

**Comment by willgearty** [comment]
It'd also be good to have a better way to access the answers to such custom fields across all classes. Right now, the only place the custom field responses are shown is on the manage page:
https://github.com/learning-unlimited/ESP-Website/blob/0a849d3424f5aab7c9c9aed5bb89a6af161c039a/esp/templates/program/modules/adminclass/manageclass.html#L59-L70

A good first step would be to move this (and maybe other fields) into `manageclass_class_info.html`, so at least the responses are accessible in the class search module. It would also be great to have a printable that has the responses to the custom fields for all classes.

**Comment by tarsoliyaansh** [comment]
Hello @willgearty 
I want to ask that is this open for contribution?

**Comment by willgearty** [comment]
@tarsoliyaansh Thanks for your interest in contributing to our project! I've assigned you to this issue. Please make sure to check out our [instructions for getting started](https://github.com/learning-unlimited/ESP-Website/wiki#getting-started) and follow our [contribution rules](https://github.com/learning-unlimited/ESP-Website/wiki#rules). Let me know if you have any questions.

**Comment by willgearty** [comment]
Just realized this is part of our GSoC projects list, so unassigning @tarsoliyaansh. Please submit a GSoC proposal if you would like to work on this issue.

**Comment by tarsoliyaansh** [comment]
Thanks for the clarification!
I’ve already set up the project locally and have been working on this issue — the implementation is almost complete from my side. Since this is part of the GSoC scope, I’m definitely planning to submit a detailed proposal around it.
Would you prefer that I open a draft PR for early feedback, or should I wait and include this work in the formal GSoC proposal? Happy to align with the expected process.

**Comment by willgearty** [comment]
It's up to you. We have a lot of PRs to go through, so no guarantees you'd get feedback on a draft PR terribly soon. Happy to answer questions or give general proposal feedback, though.

**Comment by DarshanCodes09** [comment]
Hi, @willgearty 
I would like to work on this issue. 
I have started exploring the codebase and understanding how teacher custom form data is currently handled. 
I would like to implement a flexible and accessible solution as suggested. 
Could you please assign this issue to me?

**Comment by willgearty** [comment]
@DarshanCodes09 This issue is part of a GSoC project and should not be worked on until the GSoC summer internships begin. If you'd like to work on it, please submit a GSoC proposal. In the meantime, please find another issue to work on.

**Comment by DarshanCodes09** [comment]
Thank you for the clarification. I understand this issue is reserved for the GSoC project period. I’ll look for other issues to contribute to for now and will consider submitting a proposal for GSoC. Appreciate the guidance!

**Comment by chandrasingh01** [comment]
Hello Developer ! @willgearty 
I have solve the issue #2707 
This is  my PR Link - https://github.com/learning-unlimited/ESP-Website/pull/4526
I would kindly request you to review my PR once and provide your valuable suggestions and feedback.

I have a question - application form and  Proposal  form  both are same which we have to submit ?

**Comment by willgearty** [comment]
@chandrasingh01 This issue is related to a GSoC project and is not accepting pull requests at this time. Please make sure to only submit PRs for issues that are assigned to you in the future. And yes, the proposal will be part of your GSoC application.

**Comment by SHREESHANTH99** [comment]
Hey @willgearty and @milescalabresi 
I've been going through the code under the "teacherreg_custom_forms" tag and the form classes like StanfordTeacherQuestionsForm and BCTeacherQuestionsForm

Currently, custom fields are hardcoded as part of a Django form class, and responses are stored as class custom_form_data without a proper schema

My proposal is as follows:
- Create a CustomFormField with attributes like field_name, label, field_type, required, choices, help_text, order, program, form_type.
- Create a CustomFormFieldResponse with a ForeignKey to ClassSubject, ForeignKey to CustomFormField, and a response_value as a JSONField.
- Dynamically generate forms from the database and store them in the cache per program.
- The existing hardcoded forms will be retained for backward compatibility

My initial plan is as follows:
- Database-based field definitions with admin support
- Dynamic form generation
- Proper schema for responses
- Move display from manageclass.html to manageclass_class_info.html as discussed

Is this aligned with the original intent of issue #2707? I would appreciate your feedback before I write my GSoC proposal

**Comment by NineNightMeow** [comment]
Hi! @willgearty May I have a brief understanding of the approval criteria for GSOC about this project？(as a student)

**Comment by willgearty** [comment]
@NineNightMeow see the [discussion section](https://github.com/learning-unlimited/ESP-Website/discussions) for plenty of answers to this question.

**Comment by willgearty** [comment]
@SHREESHANTH99 yes, that sounds reasonable. Keep in mind that this issue is part of a broader GSoC project about form accessibility.

**Comment by SHREESHANTH99** [comment]
@willgearty Thank you for the confirmation! I've read through the broader context and understand this project covers form accessibility as a whole — including both the navigation linking issue (#3690) and the flexible custom fields issue (#2707).

My full proposal will address both:
1.  A FormConfigRegistry + ConfigurableFormMixin to add contextual "configure this form" links visible to admins (addressing #3690)

2.  The CustomFormField / CustomFormFieldResponse DB-backed dynamic fields system with a Django admin interface (addressing #2707)

3.Exposing custom response data in class search, printables, and CSV exports

I'm currently working on the GSoC proposal. Would it be helpful to share a draft for early feedback before the submission deadline?

**Comment by willgearty** [comment]
In the interest of time, we can't give line-by-line edits, but we're happy to give a quick look over your proposal, You can email it to the mentors list at gsoc-mentors-2026@learningu.org.

**Comment by Saurabhtbj1201** [comment]
Hi, @willgearty
I am a GSoC 2026 applicant interested in contributing to Learning Unlimited.

I would like to work on this issue (#2707) related to improving custom field flexibility in teacher registration forms. I have started exploring the codebase and understanding how form customization is currently implemented.

Could you please assign this issue to me? I would appreciate any guidance or suggestions before I begin.
Thank you!

**Comment by khushpreet78141** [comment]
Hi @willgearty and @milescalabresi,

I’m a GSoC 2026 applicant currently exploring the ESP codebase, specifically the teacher registration custom forms and related discussions around issues #2707 and #3690.

From my understanding, the goal is not just dynamic field flexibility but also improving overall form accessibility — including better navigation to configuration points and reducing reliance on hardcoded Django forms.

I’m planning to propose a solution involving:
- A database-driven CustomFormField system for flexible field definitions
- Dynamic form generation using Django forms
- Admin-accessible configuration interfaces
- Integration with exports and printables

Before finalizing my proposal, I wanted to clarify one thing:
Would you prefer responses to be stored in a structured relational format for better querying, or is a JSON-based flexible schema acceptable given the variability across chapters?

I want to ensure my design aligns well with long-term maintainability and ESP’s existing data workflows.

Thanks for your time!

**Comment by willgearty** [comment]
@khushpreet78141 That's really up for you to identify in your proposal. It's a proposal, you should be proposing what you think is best.

**Comment by khushpreet78141** [comment]
> [@khushpreet78141](https://github.com/khushpreet78141) That's really up for you to identify in your proposal. It's a proposal, you should be proposing what you think is best.

Thanks for the clarification!

I’ll make sure to clearly justify the design choice in my proposal by evaluating both relational and JSON-based approaches in terms of scalability, queryability, and flexibility for ESP’s use cases.

Appreciate the guidance!

**Comment by khushpreet78141** [comment]
Hi @willgearty,

I’ve been exploring this issue as part of understanding the broader goal of improving accessibility of custom form fields.

As an initial step, I submitted a PR (#5688) where I moved the rendering of `class.custom_form_data` into the shared `manageclass_class_info.html` template. This helps make custom field responses accessible across multiple views instead of being limited to the manage page.

I understand this issue is part of a larger GSoC project, so this change is intended as a small incremental improvement toward better accessibility of custom field data.

I’d really appreciate any feedback on whether this approach aligns with the intended direction.

Thank you!

**Comment by willgearty** [comment]
@khushpreet78141 It is very clear in the discussion above that no one should be working on this issue because it is part of a GSoC project. If you are interested in contributing, there are many other unassigned issues. Furthermore, our contributor rules are very clear that you should not submit a PR until you are assigned to an issue.

**Comment by khushpreet78141** [comment]
Hi @willgearty,
Thanks for pointing this out. I apologize for missing the discussion and the contribution guidelines regarding issue assignment.
I understand now that this issue is part of a GSoC project and shouldn’t have been worked on.
I’ll make sure to follow the proper process going forward and will look for other unassigned issues to contribute to.
Thanks for your guidance!

**Comment by RITVIKKAMASETTY** [comment]
Hi @willgearty,

I had submitted a proposal for this project for GSoC, but it wasn’t selected. However, I’ve already done a good amount of research and groundwork toward implementing it. I’d really love to contribute to this issue would it be possible for me to be assigned to it?

Thanks!
